File: day2/part2.hs
import Data.List (unfoldr, splitAt)
import Data.Bool (bool)
import Data.Maybe (fromMaybe)

safeTail :: [a] -> Maybe [a]
safeTail = (bool Nothing . (Just . tail)) <*> (not . null)

splitOn :: Eq a => a -> [a] -> [[a]]
splitOn c = unfoldr go where
    go s = if null s then Nothing else Just $ split s
    split s = fromMaybe [] . safeTail <$> break (== c) s

parse :: String -> [(Int, Int)]
parse = map parse' . splitOn ',' where
    parse' line = case splitOn '-' line of
        [a, b] -> (read a, read b)
        _ -> undefined

isInvalid :: Int -> Bool
isInvalid num = match where
    s = show num
    n = length s
    ixs = [1..(n `div` 2)]
    isDiv a b = a `mod` b == 0
    jxs = filter (n `isDiv`) ixs
    rep i = concat $ replicate (n `div` i) (take i s)
    match = any ((s ==) . rep) jxs

solve :: [(Int, Int)] -> Int
solve = sum . filter isInvalid . nums where
    enumFromTo' = uncurry enumFromTo
    nums = concatMap enumFromTo'

main :: IO ()
main = interact (show . solve . parse)

File: day2/part2.py
from sys import stdin
from typing import List, Tuple

def is_invalid(num):
    s = str(num)
    for i in range(1, len(s) // 2 + 1):
        if len(s) % i != 0:
            continue
        match = (
            s[j*i:(j+1)*i] == s[:i]
            for j in range(1, len(s) // i))
        if all(match):
            return True
    return False

def solve(ranges: List[Tuple[int, int]]) -> int:
    total = 0
    for a, b in ranges:
        for i in range(a, b + 1):
            if is_invalid(i):
                total += i
    return total

def parse(s: str) -> List[Tuple[int, int]]:
    out = []
    for s1 in s.split(','):
        a, b = s1.split('-')
        a, b = int(a), int(b)
        out.append((a, b))
    return out

if __name__ == '__main__':
    print(solve(parse(stdin.read())))